Source/WebKit/GPUProcess/graphics/WebGPU/WebGPUObjectHeap.cpp

```diff
@@ -53,10 +53,119 @@
+            return "RemoteExternalTexture";
+        },
+        [] (const IPC::ScopedActiveMessageReceiveQueue<RemotePipelineLayout>& pipelineLayout) {
+            return "RemotePipelineLayout";
+        },
+        [] (const IPC::ScopedActiveMessageReceiveQueue<RemoteQuerySet>& querySet) {
+            return "RemoteQuerySet";
+        },
+        [] (const IPC::ScopedActiveMessageReceiveQueue<RemoteQueue>& queue) {
+            return "RemoteQueue";
+        },
+        [] (const IPC::ScopedActiveMessageReceiveQueue<RemoteRenderBundleEncoder>& renderBundleEncoder) {
+            return "RemoteRenderBundleEncoder";
+        },
+        [] (const IPC::ScopedActiveMessageReceiveQueue<RemoteRenderBundle>& adapter) {
+            return "RemoteRenderBundle";
+        },
+        [] (const IPC::ScopedActiveMessageReceiveQueue<RemoteRenderPassEncoder>& bindGroup) {
+            return "RemoteRenderPassEncoder";
+        },
+        [] (const IPC::ScopedActiveMessageReceiveQueue<RemoteRenderPipeline>& bindGroupLayout) {
+            return "RemoteRenderPipeline";
+        },
+        [] (const IPC::ScopedActiveMessageReceiveQueue<RemoteSampler>& adapter) {
+            return "RemoteSampler";
+        },
+        [] (const IPC::ScopedActiveMessageReceiveQueue<RemoteShaderModule>& bindGroup) {
+            return "RemoteShaderModule";
+        },
+        [] (const IPC::ScopedActiveMessageReceiveQueue<RemoteSurface>& bindGroupLayout) {
+            return "RemoteSurface";
+        },
+        [] (const IPC::ScopedActiveMessageReceiveQueue<RemoteSwapChain>& adapter) {
+            return "RemoteSwapChain";
+        },
+        [] (const IPC::ScopedActiveMessageReceiveQueue<RemoteTexture>& bindGroup) {
+            return "RemoteTexture";
+        },
+        [] (const IPC::ScopedActiveMessageReceiveQueue<RemoteTextureView>& bindGroupLayout) {
+            return "RemoteTextureView";
+        },
+        [] (auto&& autoObject) {
+            (void)autoObject;
+            return "empty";
+        });
+}
+void ObjectHeap::printObjectHeap()
+{
+    std::unordered_map<char const*, int> counts;
+    for (auto& keyValuePair : m_objects)
+        ++counts[typeOfObject(keyValuePair.value)];
+
+    WTFLogAlways("===================== WebGPU ObjectHeap table =====================");
+    for (auto& keyValuePair : counts)
+        WTFLogAlways("%s - %d", keyValuePair.first, keyValuePair.second);
+
+    WTFLogAlways("===================================================================\n");
+}
+#endif // PLATFORM(COCOA)
+
 ObjectHeap::ObjectHeap()
 {
+#if PLATFORM(COCOA)
+    static ObjectHeap* currentObjectHeap;
+    currentObjectHeap = this;
+    // "notifyutil -p com.apple.WebKit.WebGPU.PrintObjectHeap"
+    static std::once_flag onceFlag;
+    std::call_once(onceFlag, [] {
+        int printObjectHeapToken;
+        notify_register_dispatch("com.apple.WebKit.WebGPU.PrintObjectHeap", &printObjectHeapToken, dispatch_get_main_queue(), ^(int) {
```

Comment:
We should be able to pass C++ lambdas to functions that accept blocks. That's better in our situation, because we can put stuff in the capture clause.

```
notify_register_dispatch("com.apple.WebKit.WebGPU.PrintObjectHeap", &printObjectHeapToken, dispatch_get_main_queue(), [objectHeap = Ref { *this }] {
    objectHeap->printObjectHeap();
});
```

(This technique may require `#include <wtf/BlockPtr.h>` too, not sure.)

I also think we shouldn't have to deal with all this `call_once` stuff, and just have every object heap that exists register for printing. If you have 2 alive, you probably want both to be dumped, instead of whichever was created second.

Reply:
Good point I will add the token as a member variable and cancel the registration in the destructor.

I don't think we should capture `*this` in a `Ref<>` because then the `notify_register_dispatch` will retain the object and keep the object alive right? We don't want it retained by `notify_register_dispatch`